packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1157,127 +1157,247 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
     MacOSPlugin.kConfigKey,
     WindowsPlugin.kConfigKey,
   ];
-  final List<PluginInterfaceResolution> finalResolution = <PluginInterfaceResolution>[];
+  final List<PluginInterfaceResolution> pluginResolutions = <PluginInterfaceResolution>[];
   bool hasResolutionError = false;
+  bool hasPluginPubspecError = false;
 
   for (final String platformKey in platformKeys) {
-    // Key: the plugin name
-    final Map<String, List<Plugin>> possibleResolutions = <String, List<Plugin>>{};
+    // Key: the plugin name, value: the list of plugin candidates for the implementation of [platformKey].
+    final Map<String, List<Plugin>> pluginImplCandidates = <String, List<Plugin>>{};
+
+    // Key: the plugin name, value: the plugin name of the default implementation of [platformKey].
     final Map<String, String> defaultImplementations = <String, String>{};
 
     for (final Plugin plugin in plugins) {
-      final String? defaultImplementation = plugin.defaultPackagePlatforms[platformKey];
-      if (plugin.platforms[platformKey] == null && defaultImplementation == null) {
-        // The plugin doesn't implement this platform.
-        continue;
-      }
-      String? implementsPackage = plugin.implementsPackage;
-      if (implementsPackage == null || implementsPackage.isEmpty) {
-        final bool hasInlineDartImplementation =
-            plugin.pluginDartClassPlatforms[platformKey] != null;
-        if (defaultImplementation == null && !hasInlineDartImplementation) {
-          // Skip native inline PluginPlatform implementation
-          continue;
+      final String? error = _validatePlugin(plugin, platformKey);
+      if (error != null) {
+        globals.printError(error);
+        hasPluginPubspecError = true;
+      } else {
```

Comment:
Consider using a `continue;` in the `if` instead of nesting all of the expected codepath an extra time.

Reply:
Done 5a2657b4f1e522d07cac57aa8086858a1369db93 .